passphrase.go

```diff
@@ -30,3 +88,193 @@ func GenerateKeyFromPassphraseAndCopy(passphrase string, key []byte) {
+	crypto.RegisterHash(crypto.SHA256, sha256.New)
+}
+
+// NewKeyValidator returns a key validator which be used either globally or in its own thread.
+// length indicates how much of the key checksum hexadecimal encoding is used for validation (0 < length <= 2*cryptoHash.Size())
+// flags: nonce | pre-allocated validator
+func NewKeyValidator(cryptoHash crypto.Hash, key []byte, length int, flags ...bool) (Validator, error) {
+	var (
+		nonce     uint32    = 0
+		nonceFlag bool      = false
+		mac       hash.Hash = nil
+		err       error     = nil
+	)
+	registerHashFunctions()
+	switch len(flags) {
+	case 0:
+		// nonce not used, on-the-fly validator
+		nonceFlag = false
+		mac = nil
+	case 1:
+		// nonce flag specified, on-the-fly validator
+		nonceFlag = flags[0]
+		mac = nil
+	case 2:
+		// nonce flag specified, pre-allocated validator flag specified
+		fallthrough
+	default:
+		nonceFlag = flags[0]
+		if flags[1] {
+			// pre-allocated validator
+			mac = hmac.New(cryptoHash.New, key)
+		} else {
+			// on-the-fly validator
+			mac = nil
+		}
+	}
+	if nonceFlag {
+		// generate a random nonce
+		if nonce, err = randomUint32(); err != nil {
+			return nil, err
+		}
+	}
+	if length < ChecksumMinLength {
+		length = ChecksumMinLength
+	} else if length > 2*cryptoHash.Size() {
+		length = 2 * cryptoHash.Size()
+	}
+	return &KeyValidator{key, cryptoHash, mac, length, nonceFlag, nonce}, nil
+}
+
+// computeWithNonce computes the validation code using an optional nonce specified as parameter
+func (vtor *KeyValidator) computeWithNonce(nonce ...uint32) (kv KeyValidation) {
+	mac := vtor.mac
+	if mac == nil {
+		// allocate an "ephemeral" hmac object
+		mac = hmac.New(vtor.hash.New, vtor.key)
+	}
+	mac.Reset()
+	mac.Write(vtor.key)
+	if len(nonce) > 0 {
+		var b [4]byte
+		kv.nonce = nonce[0]
+		kv.withNonce = true
+		// use the nonce in big endian format
+		binary.BigEndian.PutUint32(b[0:], kv.nonce)
+		// append it to the buffer over which the hash is computed
+		mac.Write(b[:])
+	}
+	kv.code = mac.Sum(nil)[:]
+	kv.length = vtor.length
+	return
+}
+
+// compute computes the key validation
+func (vtor *KeyValidator) compute() (kv KeyValidation) {
+	if vtor.withNonce {
+		return vtor.computeWithNonce(atomic.AddUint32(&vtor.nonce, 1))
+	}
+	return vtor.computeWithNonce()
+}
+
+// Compute computes the key validation and returns its string representation
+func (vtor *KeyValidator) Compute() (code string) {
+	kv := vtor.compute()
+	return kv.String()
+}
+
+func (vtor *KeyValidator) String() string {
+	if vtor.withNonce {
+		return fmt.Sprintf("%v:%s:%d", vtor.key, vtor.hash, vtor.nonce)
+	}
+	return fmt.Sprintf("%v:%s", vtor.key, vtor.hash)
+}
+
+func (vtor *KeyValidator) Validate(code string) (isValid bool) {
+	var kv KeyValidation
+	if n, hasNonce := getNonce(code); hasNonce {
+		kv = vtor.computeWithNonce(n)
+	} else {
+		kv = vtor.computeWithNonce()
+	}
+	isValid = (subtle.ConstantTimeCompare([]byte(code), []byte(kv.String())) == 1)
```

Comment:
I would add a ValidateBytes(code []byte), just to avoid the []byte(string) conversion in the case when we get []byte (which would be our case if we ever validate something). The []byte(string) would create a copy and most likely the copy will escape and be allocated on the heap (you could avoid it via unsafe, but I think is better to have a ValidateBytes() option).

Reply:
we do not validate on the probe side, right? we compute the validation code on the probe and we validate on the gui / js type. I only added this for the unit tests.